spring-web-modules/spring-rest-http-3/src/main/java/com/baeldung/bulkandbatchapi/service/CustomerService.java

```diff
@@ -0,0 +1,105 @@
+package com.baeldung.bulkandbatchapi.service;
+
+import com.baeldung.bulkandbatchapi.request.Customer;
+import com.baeldung.bulkandbatchapi.request.BulkActionType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import java.util.*;
+
+@Service
+public class CustomerService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CustomerService.class);
+    private final Map<String, Customer> customerRepoMap = new HashMap<>();
+
+    public List<Customer> createCustomers(List<Customer> customers) {
+        List<Customer> createdCustomers = new ArrayList<>();
+
+        customers.forEach(customer ->
+        {
+            Optional<Customer> customerCreated = createCustomer(customer);
+            customerCreated.ifPresent(createdCustomers::add);
+        });
+
+        return createdCustomers;
+    }
+
+    public List<Customer> processCustomers(List<Customer> customers, BulkActionType batchType) {
+        List<Customer> processedCustomers = new ArrayList<>();
+
+        customers.forEach(customer ->
+        {
+            Optional<Customer> customerProcessed = Optional.empty();
+            switch (batchType) {
+                case CREATE:
+                    customerProcessed = createCustomer(customer);
+                    break;
+                case UPDATE:
+                    customerProcessed = updateCustomer(customer);
+                    break;
+                case DELETE:
+                    customerProcessed = deleteCustomer(customer);
+                    break;
+                default:
+                    break;
+            }
+
+            customerProcessed.ifPresent(processedCustomers::add);
+        });
```

Comment:
This could also be done as a stream/collect.

However, the big body that's in L-34-47 here should be extracted into a separate method. Arguable, the operation could be detected outside of the loop and held as a `Function<Customer, Customer>`

Reply:
Yes, can be stream/collect. Now, i've defined this as BiFunction outside the stream.